agent/grpc-external/limiter/limiter.go

```diff
@@ -0,0 +1,245 @@
+// package limiter provides primatives for limiting the number of concurrent
+// operations in-flight.
+package limiter
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"sync/atomic"
+
+	"golang.org/x/time/rate"
+)
+
+// Unlimited can be used to allow an unlimited number of concurrent sessions.
+const Unlimited uint32 = 0
+
+// ErrCapacityReached is returned when there is no capacity for additional sessions.
+var ErrCapacityReached = errors.New("active session limit reached")
+
+// SessionLimiter is a session-based concurrency limiter, it provides the basis
+// of gRPC/xDS load balancing.
+//
+// Stream handlers obtain a session with BeginSession before they begin serving
+// resources - if the server has reached capacity ErrCapacityReached is returned,
+// otherwise a Session is returned.
+//
+// It is the session-holder's responsibility to:
+//
+//	1. Call End on the session when finished.
+//	2. Receive on the session's Terminated channel and exit (e.g. close the gRPC
+//	   stream) when it is closed.
+//
+// The maximum number of concurrent sessions is controlled with SetMaxSessions.
+// If there are more than the given maximum sessions already in-flight,
+// SessionLimiter will drain randomly-selected sessions at a rate controlled
+// by SetDrainRateLimit.
+type SessionLimiter struct {
```

Comment:
Note: we refer to xDS sessions as "streams" in most places for consistency with existing metrics etc. but this component is called the "session limiter" because it is generic and can be used to limit any concurrent operations.